File: src/microservices/_old_c#/events/EventsController.cs
using EventsService.Kafka;
using EventsService.Models;
using Microsoft.AspNetCore.Mvc;

namespace EventsService.Controllers
{
    [ApiController]
    [Route("api/events")]
    public class EventsController : ControllerBase
    {
        private readonly EventProducer _producer;
        private readonly ILogger<EventsController> _logger;

        public EventsController(EventProducer producer, ILogger<EventsController> logger)
        {
            _producer = producer;
            _logger = logger;
        }
        
        // GET /api/events/health
        [HttpGet("health")]
        [ProducesResponseType(typeof(object), 200)]
        public IActionResult GetHealth()
        {
            _logger.LogInformation("Health check requested.");
            return Ok(new { status = true });
        }

        // POST /api/events/movie
        [HttpPost("movie")]
        [ProducesResponseType(typeof(EventResponse), 201)]
        [ProducesResponseType(500)]
        public async Task<IActionResult> CreateMovieEvent([FromBody] MovieEvent movieEvent)
        {
            try
            {
                var result = await _producer.ProduceAsync("movie-events", movieEvent);
                return StatusCode(201, result);
            }
            catch (Exception ex)
            {
                _logger.LogError(ex, "Failed to produce movie event to Kafka.");
                return StatusCode(500, new { error = "Internal Server Error during movie event production." });
            }
        }

        // POST /api/events/user
        [HttpPost("user")]
        [ProducesResponseType(typeof(EventResponse), 201)]
        [ProducesResponseType(500)]
        public async Task<IActionResult> CreateUserEvent([FromBody] UserEvent userEvent)
        {
            try
            {
                var result = await _producer.ProduceAsync("user-events", userEvent);
                return StatusCode(201, result);
            }
            catch (Exception ex)
            {
                _logger.LogError(ex, "Failed to produce user event to Kafka.");
                return StatusCode(500, new { error = "Internal Server Error during user event production." });
            }
        }

        // POST /api/events/payment
        [HttpPost("payment")]
        [ProducesResponseType(typeof(EventResponse), 201)]
        [ProducesResponseType(500)]
        public async Task<IActionResult> CreatePaymentEvent([FromBody] PaymentEvent paymentEvent)
        {
            try
            {
                var result = await _producer.ProduceAsync("payment-events", paymentEvent);
                return StatusCode(201, result);
            }
            catch (Exception ex)
            {
                _logger.LogError(ex, "Failed to produce payment event to Kafka.");
                return StatusCode(500, new { error = "Internal Server Error during payment event production." });
            }
        }
    }
}

File: src/microservices/events/internal/api/events_controller.go
package api

import (
	"encoding/json"
	"events-service/internal/kafka"
	"events-service/internal/models"
	"log"
	"net/http"

	"github.com/go-chi/render"
)

// EventsHandler - Контроллер, который содержит зависимости (EventProducer)
type EventsHandler struct {
	producer *kafka.EventProducer
}

// NewEventsHandler - Конструктор с инъекцией зависимостей
func NewEventsHandler(p *kafka.EventProducer) *EventsHandler {
	return &EventsHandler{producer: p}
}

// writeJSONResponse - Хелпер для отправки JSON-ответа
func writeJSONResponse(w http.ResponseWriter, statusCode int, data interface{}) {
	w.Header().Set("Content-Type", "application/json")
	w.WriteHeader(statusCode)
	if err := json.NewEncoder(w).Encode(data); err != nil {
		log.Printf("Error writing JSON response: %v", err)
	}
}

// GetHealth - GET /api/events/health
func (h *EventsHandler) GetHealth(w http.ResponseWriter, r *http.Request) {
	log.Println("Health check requested.")
	writeJSONResponse(w, http.StatusOK, map[string]bool{"status": true})
}

// handleEventCreation - Универсальная функция для POST-запросов
func (h *EventsHandler) handleEventCreation(w http.ResponseWriter, r *http.Request, topic string, eventModel models.Eventer) {
	// 1. Десериализация
	if err := json.NewDecoder(r.Body).Decode(eventModel); err != nil {
		http.Error(w, "Invalid request body", http.StatusBadRequest)
		return
	}

	// 2. Инициализация полей BaseEvent
	// Это критический шаг: в C# он был в конструкторе/свойствах. 
	// В Go это нужно делать явно, встраивая BaseEvent в структуру и присваивая новые значения.
	baseEvent := models.NewBaseEvent()
	// Используем reflect или интерфейс, чтобы обновить встроенные поля BaseEvent
	// Это обходной путь. Более чистый способ - создать NewXEvent() функции, 
	// которые инициализируют BaseEvent. Но для соответствия C# логике:
	switch v := eventModel.(type) {
	case *models.MovieEvent:
		v.BaseEvent = baseEvent
	case *models.UserEvent:
		v.BaseEvent = baseEvent
	case *models.PaymentEvent:
		v.BaseEvent = baseEvent
	}

	// 3. Отправка в Kafka
	result, err := h.producer.ProduceAsync(topic, eventModel)
	if err != nil {
		log.Printf("Failed to produce event to Kafka for topic %s: %v", topic, err)
		writeJSONResponse(w, http.StatusInternalServerError, map[string]string{
			"error": "Internal Server Error during event production.",
		})
		return
	}

	// 4. Успешный ответ
	writeJSONResponse(w, http.StatusCreated, result)
}

// CreateMovieEvent - POST /api/events/movie
func (h *EventsHandler) CreateMovieEvent(w http.ResponseWriter, r *http.Request) {
	h.handleEventCreation(w, r, "movie-events", &models.MovieEvent{})
}

// CreateUserEvent - POST /api/events/user
func (h *EventsHandler) CreateUserEvent(w http.ResponseWriter, r *http.Request) {
	h.handleEventCreation(w, r, "user-events", &models.UserEvent{})
}

// CreatePaymentEvent - POST /api/events/payment
func (h *EventsHandler) CreatePaymentEvent(w http.ResponseWriter, r *http.Request) {
	h.handleEventCreation(w, r, "payment-events", &models.PaymentEvent{})
}